File: demo/KSYLiveDemo/KSYPlayerDemo/KSYProberVC.m
//
//  KSYProberVC.m
//  KSYPlayerDemo
//
//  Created by 施雪梅 on 16/7/10.
//  Copyright © 2016年 kingsoft. All rights reserved.
//
#import "KSYUIView.h"
#import "KSYProberVC.h"

#define ELEMENT_GAP  10

@implementation KSYProberVC {
    KSYUIView *ctrlView;
    UILabel *stat;
    UIButton *btnProbe;//格式探测
    UIButton *btnThumbnail;//保存缩略图
    UIButton *btnQuit;//退出
    
    NSURL *_url;
    KSYMediaInfoProber *_prober;
}

- (instancetype)initWithURL:(NSURL *)url {
    if((self = [super init])) {
        _url = url;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupUI];
    
    stat.text = [NSString stringWithFormat:@"url is : %@", [_url isFileURL] ? [_url path] : [_url absoluteString]];
    _prober = [[KSYMediaInfoProber alloc] initWithContentURL: _url];
    _prober.timeout = 10;
}

- (void)setupUI {
    ctrlView = [[KSYUIView alloc] initWithFrame:self.view.bounds];
    ctrlView.backgroundColor = [UIColor whiteColor];
    //设置按钮与按钮间的间隔
    ctrlView.gap = ELEMENT_GAP;     
    
    @WeakObj(self);
    ctrlView.onBtnBlock = ^(id sender){
        [selfWeak  onBtn:sender];
    };
    
    btnProbe = [ctrlView addButton:@"探测格式"];
    btnThumbnail = [ctrlView addButton:@"缩略图"];
    btnQuit = [ctrlView addButton:@"退出"];

    stat = [ctrlView addLable:nil];
    stat.backgroundColor = [UIColor clearColor];
    stat.textColor = [UIColor redColor];
    stat.numberOfLines = -1;
    stat.textAlignment = NSTextAlignmentLeft;
    
    [self layoutUI];
    
    [self.view addSubview: ctrlView];
}

- (void)layoutUI {
    ctrlView.frame = self.view.frame;
    [ctrlView layoutUI];
    
    ctrlView.yPos  = ctrlView.frame.size.height -  ctrlView.btnH - ELEMENT_GAP;
    [ctrlView putRow:@[btnProbe, btnThumbnail, btnQuit]];

    stat.frame = ctrlView.frame;
}

-(BOOL)textFieldShouldReturn:(UITextField *)textField
{
    //收回键盘
    if (textField.returnKeyType == UIReturnKeyDone) {
        [textField resignFirstResponder];
    }
    return YES;
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
}

- (void)onBtn:(UIButton *)btn{
    if (btn == btnProbe) {
        //格式探测
        [self onProbeMediaInfo];
    }else if (btn == btnThumbnail){
        //保存缩略图
        [self onThumbnail];
    }else if (btn == btnQuit){
        //退出格式探测
        [self onQuit];
    }
}

- (void)onProbeMediaInfo {

    if(nil == _prober)
        return ;
    //显示探测结果
    NSMutableString *result = [[NSMutableString alloc] init];
    int i = 0;
    //媒体信息
    KSYMediaInfo *mediaInfo = _prober.ksyMediaInfo;
    if(mediaInfo)
    {
        //媒体的封装格式
        [result appendFormat:@"\nmux type:%@", [self convertMuxType:mediaInfo.type]];
        //媒体的码率
        [result appendFormat:@"\nbitrate:%lld", mediaInfo.bitrate];
        
        i = 0;
        //视频流的数量
        [result appendFormat:@"\n\nvideo num is : %lu", (unsigned long)[mediaInfo.videos count]];
        for (KSYVideoInfo  *videoInfo in mediaInfo.videos) {
            //各个视频流类型及视频帧的宽度和高度
            [result appendFormat:@"\n\nvideo[%d] codec:%@", i, [self convertAVCodec:videoInfo.vcodec]];
            [result appendFormat:@"\nvideo[%d] frame width:%d", i, videoInfo.frame_width];
            [result appendFormat:@"\nvideo[%d] frame height:%d", i, videoInfo.frame_height];
            i++;
        }
        
        i = 0;
        //音频流的数量
        [result appendFormat:@"\n\naudio num is : %lu", (unsigned long)[mediaInfo.audios count]];
        for(KSYAudioInfo  *audioInfo in mediaInfo.audios)
        {
            //各个音频流的信息，包含类型、语言、码率、声道、音频帧大小、FMT、采样率
            [result appendFormat:@"\n\naudio[%d] codec:%@", i, [self convertAVCodec:audioInfo.acodec]];
            [result appendFormat:@"\naudio[%d] language:%@", i, audioInfo.language];
            [result appendFormat:@"\naudio[%d] bitrate:%lld", i, audioInfo.bitrate];
            [result appendFormat:@"\naudio[%d] channels:%d", i, audioInfo.channels];
            [result appendFormat:@"\naudio[%d] frame_size:%d", i, audioInfo.framesize];
            [result appendFormat:@"\naudio[%d] sample_format:%@", i, [self convertSampleFMT:audioInfo.sample_format]];
            [result appendFormat:@"\naudio[%d] samplerate:%d", i, audioInfo.samplerate];
            i++;
        }
    }
    else
        //文件探测失败
        [result appendFormat:@"\nprobe mediainfo failed!"];
    
    stat.text = [NSString stringWithFormat:@"%@", result];
}

- (void)onThumbnail{
    
    if(nil == _prober)
        return ;
    //获取视频缩略图
    UIImage *thumbnailImage = [_prober getVideoThumbnailImageAtTime:0 width:0 height:0];
    if(thumbnailImage)
        //保存缩略图
        [KSYUIVC saveImageToPhotosAlbum:thumbnailImage];
    else
    {
        //截图失败
        UIAlertView *toast = [[UIAlertView alloc] initWithTitle:@"￣へ￣"
                                                        message:@"缩略图截取失败！"
                                                       delegate:nil
                                              cancelButtonTitle:@"确定"
                                              otherButtonTitles:nil, nil];
        [toast show];
    }
}

- (void)onQuit {
    [self dismissViewControllerAnimated:FALSE completion:nil];
}

#pragma mark converMediaInfo

- (NSString *)convertMuxType:(MEDIAINFO_MUX_TYPE)muxType
{
    //返回视频的封装格式
    NSString *muxTypeStr = @"unknow mux type";

    if(MEDIAINFO_MUXTYPE_MP2T == muxType)
        muxTypeStr = @"mpeg-ts";
    else if(MEDIAINFO_MUXTYPE_MOV == muxType)
        muxTypeStr = @"mov";
    else if(MEDIAINFO_MUXTYPE_AVI == muxType)
        muxTypeStr = @"avi";
    else if(MEDIAINFO_MUXTYPE_FLV == muxType)
        muxTypeStr = @"flv";
    else if(MEDIAINFO_MUXTYPE_MKV == muxType)
        muxTypeStr = @"mkv";
    else if(MEDIAINFO_MUXTYPE_ASF == muxType)
        muxTypeStr = @"asf";
    else if(MEDIAINFO_MUXTYPE_RM == muxType)
        muxTypeStr = @"rm";
    else if(MEDIAINFO_MUXTYPE_WAV == muxType)
        muxTypeStr = @"wav";
    else if(MEDIAINFO_MUXTYPE_OGG == muxType)
        muxTypeStr = @"ogg";
    else if(MEDIAINFO_MUXTYPE_APE == muxType)
        muxTypeStr = @"ape";
    else if(MEDIAINFO_MUXTYPE_RAWVIDEO == muxType)
        muxTypeStr = @"rawvideo";
    else if(MEDIAINFO_MUXTYPE_HLS == muxType)
        muxTypeStr = @"hls";
    
    return muxTypeStr;
}

- (NSString *)convertAVCodec:(MEDIAINFO_CODEC_ID)codecID
{
    //返回音频的封装格式
    NSString *codecIDStr = @"unknow codec";
    
    if(MEDIAINFO_CODEC_MPEG2VIDEO == codecID)
        codecIDStr = @"mpeg2";
    else if(MEDIAINFO_CODEC_MPEG4 == codecID)
        codecIDStr = @"mpeg4";
    else if(MEDIAINFO_CODEC_MJPEG == codecID)
        codecIDStr = @"mjpeg";
    else if(MEDIAINFO_CODEC_JPEG2000 == codecID)
        codecIDStr = @"jpeg2000";
    else if(MEDIAINFO_CODEC_H264 == codecID)
        codecIDStr = @"h264";
    else if(MEDIAINFO_CODEC_HEVC == codecID)
        codecIDStr = @"hevc";
    else if(MEDIAINFO_CODEC_VC1 == codecID)
        codecIDStr = @"vc1";
    else if(MEDIAINFO_CODEC_AAC == codecID)
        codecIDStr = @"aac";
    else if(MEDIAINFO_CODEC_AC3 == codecID)
        codecIDStr = @"ac3";
    else if(MEDIAINFO_CODEC_MP3 == codecID)
        codecIDStr = @"mp3";
    else if(MEDIAINFO_CODEC_PCM == codecID)
        codecIDStr = @"pcm";
    else if(MEDIAINFO_CODEC_DTS == codecID)
        codecIDStr = @"dts";
    else if(MEDIAINFO_CODEC_NELLYMOSER == codecID)
        codecIDStr = @"nellymoser";
    
    return codecIDStr;
}


- (NSString *)convertSampleFMT:(MEDIAINFO_SAMPLE_FMT)afmt
{
    //返回sample的FMT
    NSString *sampleFMTStr = @"unknown sample formats";
    
    if(MEDIAINFO_SAMPLE_FMT_U8 == afmt)
        sampleFMTStr = @"unsigned 8 bits";
    else if(MEDIAINFO_SAMPLE_FMT_S16 == afmt)
        sampleFMTStr = @"signed 16 bits";
    else if(MEDIAINFO_SAMPLE_FMT_S32 == afmt)
        sampleFMTStr = @"signed 32 bits";
    else if(MEDIAINFO_SAMPLE_FMT_FLT == afmt)
        sampleFMTStr = @"float";
    else if(MEDIAINFO_SAMPLE_FMT_DBL == afmt)
        sampleFMTStr = @"double";
    else if(MEDIAINFO_SAMPLE_FMT_U8P == afmt)
        sampleFMTStr = @"unsigned 8 bits, planar";
    else if(MEDIAINFO_SAMPLE_FMT_S16P == afmt)
        sampleFMTStr = @"signed 16 bits, planar";
    else if(MEDIAINFO_SAMPLE_FMT_S32P == afmt)
        sampleFMTStr = @"signed 32 bits, planar";
    else if(MEDIAINFO_SAMPLE_FMT_FLTP == afmt)
        sampleFMTStr = @"float, planar";
    else if(MEDIAINFO_SAMPLE_FMT_DBLP == afmt)
        sampleFMTStr = @"double, planar";
    else if(MEDIAINFO_SAMPLE_FMT_NB == afmt)
        sampleFMTStr = @"Number of sample formats";

    return sampleFMTStr;
}

@end

File: demo/KSYLiveDemo_Swift/KSYPlayerDemo_Swift/KSYProberVC.swift
//
//  KSYProberVC.swift
//  KSYLiveDemo_Swift
//
//  Created by iVermisseDich on 17/1/13.
//  Copyright © 2017年 com.ksyun. All rights reserved.
//

import UIKit

class KSYProberVC: UIViewController {
    
    var url: URL?
    var prober: KSYMediaInfoProber?
    
    var stat: UILabel?
    
    var btnProbe: UIButton?
    var btnThumbnail: UIButton?
    var btnQuit: UIButton?
    
    
    init() {
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    convenience init(url: URL) {
        self.init()
        self.url = url
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        initUI()
    }
    
    func initUI() {
        view.backgroundColor = .white
        // add play button
        btnProbe = addButton(with: "probe", action: #selector(onProbeMediaInfo(sender:)))
        
        // add Thunbnail button
        btnThumbnail = addButton(with: "Thumbnail", action: #selector(onThumbnail(sender:)))
        
        // add quit button
        btnQuit = addButton(with: "quit", action: #selector(onQuit(sender:)))
        
        stat = UILabel()
        stat?.backgroundColor = .clear
        stat?.textColor = .red
        stat?.numberOfLines = -1
        stat?.textAlignment = .left
        view.addSubview(stat!)
        layoutUI()
        
        let aUrlString = url!.isFileURL ? url?.path : url?.absoluteString
        
        stat?.text = "url is : \(aUrlString ?? "")"
        prober = KSYMediaInfoProber.init(contentURL: url)
        prober?.timeout = 10
    }
    
    func addButton(with title: String, action: Selector) -> UIButton {
        let button = UIButton.init(type: .roundedRect)
        button.setTitle(title, for: .normal)
        button.backgroundColor = UIColor.lightGray
        button.addTarget(self, action: action, for: .touchUpInside)
        button.layer.masksToBounds = true
        button.layer.cornerRadius = 5
        button.layer.borderColor = UIColor.black.cgColor
        button.layer.borderWidth = 1
        view.addSubview(button)
        
        return button
    }
    
    func layoutUI() {
        let wdt = view.bounds.width
        let hgt = view.bounds.height
        
        let gap: CGFloat = 20
        let btnWdt = (wdt - gap) / 3 - gap
        let btnHgt: CGFloat = 30
        
        var xPos = gap
        let yPos = hgt - btnHgt - gap
        
        btnProbe?.frame = CGRect.init(x: xPos,
                                      y: yPos,
                                      width: btnWdt,
                                      height: btnHgt)
        
        xPos += gap + btnWdt
        btnThumbnail?.frame = CGRect.init(x: xPos,
                                          y: yPos,
                                          width: btnWdt,
                                          height: btnHgt)
        xPos += gap + btnWdt
        btnQuit?.frame = CGRect.init(x: xPos,
                                     y: yPos,
                                     width: btnWdt,
                                     height: btnHgt)
        
        stat?.frame = CGRect.init(x: 20, y: 0, width: wdt, height: hgt)
    }
    
    func onProbeMediaInfo(sender: NSObject) {
        guard let _ = prober else {
            return
        }
        
        var result: String = ""
        if let mediaInfo = prober?.ksyMediaInfo {
            result.append("\nmux type:\(convertMuxType(type: mediaInfo.type))")
            result.append(String.init(format: "\nbitrate:%lld", mediaInfo.bitrate))

            var i = 0
            result.append("\n\nvideo num is : \(mediaInfo.videos.count)")
            for videoInfo in mediaInfo.videos {
                result.append("\n\nvideo[\(i)] codec:\(convertAVCodec(codecID: (videoInfo as! KSYVideoInfo).vcodec))")
                result.append("\nvideo[\(i)] frame width:\((videoInfo as! KSYVideoInfo).frame_width)")
                result.append("\nvideo[\(i)] frame height:\((videoInfo as! KSYVideoInfo).frame_height)")
                i += 1
            }
            
            i = 0
            result.append("\n\naudio num is : \(mediaInfo.audios.count)")
            for audioInfo in mediaInfo.audios {
                result.append("n\naudio[\(i)] codec:\(convertAVCodec(codecID: (audioInfo as! KSYAudioInfo).acodec))")
                result.append("\naudio[\(i)] language:\((audioInfo as! KSYAudioInfo).language)")
                result.append("\naudio[\(i)] bitrate:\((audioInfo as! KSYAudioInfo).bitrate)")
                result.append("\naudio[\(i)] channels:\((audioInfo as! KSYAudioInfo).channels)")
                result.append("\naudio[\(i)] frame_size:\((audioInfo as! KSYAudioInfo).framesize)")
                result.append("\naudio[\(i)] sample_format:\(convertSampleFMT(afmt :(audioInfo as! KSYAudioInfo).sample_format))")
                result.append("\naudio[\(i)] samplerate:\((audioInfo as! KSYAudioInfo).samplerate)")
                i += 1
            }
        }else{
            result.append("\nprobe mediainfo failed!")
        }
        stat?.text = result
    }
    
    func onThumbnail(sender: NSObject) {
        guard let _ = prober else {
            return
        }
        
        let thumbnailImage = prober?.getVideoThumbnailImage(atTime: 0, width: 0, height: 0)
        
        if let _ = thumbnailImage {
            UIImageWriteToSavedPhotosAlbum(thumbnailImage!, self, #selector(didFinishSaving(image:error:contextInfo:)), nil)
        }else {
            let toast = UIAlertView.init(title: "￣へ￣", message: "缩略图截取失败！", delegate: nil, cancelButtonTitle: "确定")
            toast.show()
        }
    }
    
    func onQuit(sender: UIButton) {
        dismiss(animated: false, completion: nil)
    }
    
    func didFinishSaving(image: UIImage, error: NSError?, contextInfo: UnsafeMutableRawPointer) {
        if let _ = error {
            let toast = UIAlertView.init(title: "￣へ￣", message: "缩略图截取失败！", delegate: nil, cancelButtonTitle: "确定")
            toast.show()
        }else{
            let toast = UIAlertView.init(title: "O(∩_∩)O~~", message: "截图已保存至手机相册", delegate: nil, cancelButtonTitle: "确定")
            toast.show()
        }
    }
    
    // MARK: - converMediaInfo
    func convertMuxType(type: MEDIAINFO_MUX_TYPE) -> String {
        var muxTypeStr = "unknow mux type"
        
        switch type {
        case .MUXTYPE_MP2T:
            muxTypeStr = "mpeg-ts"
        case .MUXTYPE_MOV:
            muxTypeStr = "mov"
        case .MUXTYPE_AVI:
            muxTypeStr = "avi"
        case .MUXTYPE_FLV:
            muxTypeStr = "flv"
        case .MUXTYPE_MKV:
            muxTypeStr = "mkv"
        case .MUXTYPE_ASF:
            muxTypeStr = "asf"
        case .MUXTYPE_RM:
            muxTypeStr = "rm"
        case .MUXTYPE_WAV:
            muxTypeStr = "wav"
        case .MUXTYPE_OGG:
            muxTypeStr = "ogg"
        case .MUXTYPE_APE:
            muxTypeStr = "ape"
        case .MUXTYPE_RAWVIDEO:
            muxTypeStr = "rawvideo"
        case .MUXTYPE_HLS:
            muxTypeStr = "hls"
        default:
            ()
        }
        
        return muxTypeStr
    }
    
    func convertAVCodec(codecID: MEDIAINFO_CODEC_ID) -> String {
        var codecIDStr = "unknow codec"
        
        switch codecID {
        case .MPEG2VIDEO:
            codecIDStr = "mpeg2"
        case .MPEG4:
            codecIDStr = "mpeg4"
        case .MJPEG:
            codecIDStr = "mjpeg"
        case .JPEG2000:
            codecIDStr = "jpeg2000"
        case .H264:
            codecIDStr = "h264"
        case .HEVC:
            codecIDStr = "hevc"
        case .VC1:
            codecIDStr = "vc1"
        case .AAC:
            codecIDStr = "aac"
        case .AC3:
            codecIDStr = "ac3"
        case .MP3:
            codecIDStr = "mp3"
        case .PCM:
            codecIDStr = "pcm"
        case .DTS:
            codecIDStr = "dts"
        case .NELLYMOSER:
            codecIDStr = "nellymoser"
        default:
            ()
        }
        
        return codecIDStr
    }
    
    func convertSampleFMT(afmt: MEDIAINFO_SAMPLE_FMT) -> String {
        var sampleFMTStr = "unknown sample formats"
        
        switch afmt {
        case .U8:
            sampleFMTStr = "unsigned 8 bits"
            break
        case .S16:
            sampleFMTStr = "signed 16 bits"
        case .S32:
            sampleFMTStr = "signed 32 bits"
        case .FLT:
            sampleFMTStr = "float"
        case .DBL:
            sampleFMTStr = "double"
        case .U8P:
            sampleFMTStr = "unsigned 8 bits, planar"
        case .S16P:
            sampleFMTStr = "signed 16 bits, planar"
        case .S32P:
            sampleFMTStr = "signed 32 bits, planar"
        case .FLTP:
            sampleFMTStr = "float, planar"
        case .DBLP:
            sampleFMTStr = "double, planar"
        case .NB:
            sampleFMTStr = "Number of sample formats"
        default:
            ()
        }
       
        return sampleFMTStr
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

}
